# DataManim 연습문제
- url : https://www.datamanim.com/dataset/99_pandas/pandasMain.html

# 01 Getting & Knowing Data

 #### Question 1
 데이터를 로드하라. 데이터는 \t를 기준으로 구분되어 있다.

In [5]:
import pandas as pd
DataUrl = 'https://raw.githubusercontent.com/Datamanim/pandas/main/lol.csv'
df = pd.read_csv(DataUrl, sep='\t')  # 구분자를 문제에서 파악하고 'sep= ' 사용했어야 했음 
type(df)

pandas.core.frame.DataFrame

#### Question 2
데이터의 상위 5개 행을 출력하라

In [4]:
df.head(5)

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
0  3326086514  1504279457970          1949         9       1           2   
1  3229566029  1497848803862          1851         9       1           1   
2  3327363504  1504360103310          1493         9       1           2   
3  3326856598  1504348503996          1758         9       1           1   
4  3330080762  1504554410899          2094         9       1           2   

   firstTower  firstInhibitor  firstBaron  firstDragon  ...  t2_towerKills  \
0           1               1           1            1  ...              5   
1           1               1           0            1  ...              2   
2           1               1           1            2  ...              2   
3           1               1           1            1  ...              0   
4           1               1           1            1  ...              3   

   t2_inhibitorKills  t2_baronKills  t2_dragonKills  t2_riftHeraldKills  \
0                  0              0               1                   1   
1                  0              0               0                   0   
2                  0              0               1                   0   
3                  0              0               0                   0   
4                  0              0               1                   0   

   t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
0      114       67       43       16       51  
1       11       67      238       51      420  
2      157      238      121       57       28  
3      164       18      141       40       51  
4       86       11      201      122       18  

[5 rows x 61 columns]

#### Questtion 3
데이터의 행과 열의 갯수를 파악하라

In [6]:
df.shape

(51490, 61)

## Question 4
전체 칼럼을 출력하라


In [10]:
print(df.columns)  # 전체 데이터 프레임이 아니라 전체 칼럼임에 유의

Index(['gameId', 'creationTime', 'gameDuration', 'seasonId', 'winner',
       'firstBlood', 'firstTower', 'firstInhibitor', 'firstBaron',
       'firstDragon', 'firstRiftHerald', 't1_champ1id', 't1_champ1_sum1',
       't1_champ1_sum2', 't1_champ2id', 't1_champ2_sum1', 't1_champ2_sum2',
       't1_champ3id', 't1_champ3_sum1', 't1_champ3_sum2', 't1_champ4id',
       't1_champ4_sum1', 't1_champ4_sum2', 't1_champ5id', 't1_champ5_sum1',
       't1_champ5_sum2', 't1_towerKills', 't1_inhibitorKills', 't1_baronKills',
       't1_dragonKills', 't1_riftHeraldKills', 't1_ban1', 't1_ban2', 't1_ban3',
       't1_ban4', 't1_ban5', 't2_champ1id', 't2_champ1_sum1', 't2_champ1_sum2',
       't2_champ2id', 't2_champ2_sum1', 't2_champ2_sum2', 't2_champ3id',
       't2_champ3_sum1', 't2_champ3_sum2', 't2_champ4id', 't2_champ4_sum1',
       't2_champ4_sum2', 't2_champ5id', 't2_champ5_sum1', 't2_champ5_sum2',
       't2_towerKills', 't2_inhibitorKills', 't2_baronKills', 't2_dragonKills',
       't2_riftHer

#### Question 5
6번째 칼럼명을 출력하라

In [11]:
df.columns[5]

'firstBlood'

## Question 6
6번째 칼럼의 데이터 타입을 확인하라

In [16]:
df[df.columns[5]].dtype

dtype('int64')

#### Question 7 
데이터의 인덱스 구성을 확인하라

In [18]:
df.index

RangeIndex(start=0, stop=51490, step=1)

#### Question 8
6번째 칼럼의 3번째 값은 무엇인가?

In [19]:
df.iloc[2, 5]

2

## Questrion 9
데이터를 로드하라. 컬럼이 한글이기에 적절한 처리를 해줘야한다.


In [21]:
DataUrl = 'https://raw.githubusercontent.com/Datamanim/pandas/main/Jeju.csv'
df = pd.read_csv(DataUrl, encoding='euc-kr')
df

id          일자   시도명  읍면동명         거주인구       근무인구        방문인구  \
0     22448  2018-01-01   제주시   도두동    32249.987   3418.266  102709.092   
1     22449  2018-01-01   제주시   외도동   213500.997  10341.172  112692.789   
2     22450  2018-01-01   제주시  이도2동  1212382.218  96920.834  541194.481   
3     22451  2018-01-01   제주시  일도1동    33991.653   6034.253   72155.919   
4     22452  2018-01-01  서귀포시   대천동   155036.925   9403.969  150882.409   
...     ...         ...   ...   ...          ...        ...         ...   
9616  32064  2020-04-30  서귀포시   표선면   228260.005  16891.732  152832.449   
9617  32065  2020-04-30   제주시   한림읍   459959.064  26007.122  249700.419   
9618  32066  2020-04-30   제주시   도두동    28397.481   3144.895   84052.697   
9619  32067  2020-04-30  서귀포시   안덕면   348037.846  29106.286  251129.660   
9620  32068  2020-04-30   제주시    연동  1010643.372  65673.477  447622.068   

           총 유동인구   평균 속도  평균 소요 시간  평균 기온  일강수량  평균 풍속  
0      138377.345  39.556    29.167    5.0   0.0   2.50  
1      336534.958  32.900    30.900    5.0   0.0   2.50  
2     1850497.533  29.538    35.692    2.9   0.0   2.40  
3      112181.825  30.000    23.500    2.9   0.0   2.40  
4      315323.303  41.583    14.375    5.1   0.0   2.30  
...           ...     ...       ...    ...   ...    ...  
9616   397984.186  44.514    34.054   18.1   0.0   4.70  
9617   735666.605  48.609    61.377   18.0   0.0   4.35  
9618   115595.073  41.053    29.421   20.3   0.0   3.00  
9619   628273.792  46.595    49.189   17.6   0.0   3.50  
9620  1523938.917  40.863    27.765   14.1   0.0   4.80  

[9621 rows x 13 columns]

#### Questrion 10
데이터 마지막 3개 행을 출력하라

In [22]:
df.tail(3)

id          일자   시도명 읍면동명         거주인구       근무인구        방문인구  \
9618  32066  2020-04-30   제주시  도두동    28397.481   3144.895   84052.697   
9619  32067  2020-04-30  서귀포시  안덕면   348037.846  29106.286  251129.660   
9620  32068  2020-04-30   제주시   연동  1010643.372  65673.477  447622.068   

           총 유동인구   평균 속도  평균 소요 시간  평균 기온  일강수량  평균 풍속  
9618   115595.073  41.053    29.421   20.3   0.0    3.0  
9619   628273.792  46.595    49.189   17.6   0.0    3.5  
9620  1523938.917  40.863    27.765   14.1   0.0    4.8

## Questrion 11
수치형 변수를 가진 칼럼을 출력하라

In [26]:
df.select_dtypes(exclude=object).columns

Index(['id', '거주인구', '근무인구', '방문인구', '총 유동인구', '평균 속도', '평균 소요 시간', '평균 기온',
       '일강수량', '평균 풍속'],
      dtype='object')

## Questrion 12 
범주형 변수를 가진 칼럼을 출력하라

In [29]:
df.select_dtypes(include=object).columns

Index(['일자', '시도명', '읍면동명'], dtype='object')

#### Questrion 13
각 컬럼의 결측치 숫자를 파악하라

In [30]:
df.isnull().sum()

id          0
일자          0
시도명         0
읍면동명        0
거주인구        0
근무인구        0
방문인구        0
총 유동인구      0
평균 속도       0
평균 소요 시간    0
평균 기온       0
일강수량        0
평균 풍속       0
dtype: int64

#### Questrion 14
각 컬럼의 데이터 수, 데이터 타입을 한번에 확인하라

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9621 entries, 0 to 9620
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        9621 non-null   int64  
 1   일자        9621 non-null   object 
 2   시도명       9621 non-null   object 
 3   읍면동명      9621 non-null   object 
 4   거주인구      9621 non-null   float64
 5   근무인구      9621 non-null   float64
 6   방문인구      9621 non-null   float64
 7   총 유동인구    9621 non-null   float64
 8   평균 속도     9621 non-null   float64
 9   평균 소요 시간  9621 non-null   float64
 10  평균 기온     9621 non-null   float64
 11  일강수량      9621 non-null   float64
 12  평균 풍속     9621 non-null   float64
dtypes: float64(9), int64(1), object(3)
memory usage: 977.3+ KB


#### Questrion 15
각 수치형 변수의 분포(사분위, 평균, 표준편차, 최대, 최소)를 확인하라

In [33]:
df.describe()  # describe 쓰면 알아서 수치형만 보여주는 듯

id          거주인구           근무인구           방문인구        총 유동인구  \
count   9621.000000  9.621000e+03    9621.000000    9621.000000  9.621000e+03   
mean   27258.000000  3.174315e+05   35471.201510  195889.561802  5.487922e+05   
std     2777.487804  2.982079e+05   40381.214775  140706.090325  4.608802e+05   
min    22448.000000  9.305552e+03    1407.936000   11538.322000  2.225181e+04   
25%    24853.000000  9.539939e+04   12074.498000   99632.153000  2.216910e+05   
50%    27258.000000  2.221105e+05   21960.928000  152805.335000  3.866935e+05   
75%    29663.000000  4.106671e+05   40192.032000  236325.109000  6.406918e+05   
max    32068.000000  1.364504e+06  263476.965000  723459.209000  2.066484e+06   

             평균 속도     평균 소요 시간        평균 기온         일강수량        평균 풍속  
count  9621.000000  9621.000000  9621.000000  9621.000000  9621.000000  
mean     41.109084    37.215873    13.550828     6.972426     2.753171  
std       8.758631    12.993786     7.745515    27.617260     1.498538  
min      24.333000    12.667000    -9.600000     0.000000     0.000000  
25%      34.250000    27.889000     7.600000     0.000000     1.700000  
50%      39.640000    34.500000    13.400000     0.000000     2.400000  
75%      49.105000    46.176000    19.700000     1.500000     3.400000  
max     103.000000   172.200000    30.400000   587.500000    13.333000

#### Questrion 16 
거주인구 칼럼의 값들을 출력하라

In [34]:
df['거주인구']

0         32249.987
1        213500.997
2       1212382.218
3         33991.653
4        155036.925
           ...     
9616     228260.005
9617     459959.064
9618      28397.481
9619     348037.846
9620    1010643.372
Name: 거주인구, Length: 9621, dtype: float64

## Questrion 17
평균 속도 칼럼의 4분위 범위(IQR) 값을 구하여라

- IQR 구하는 공식 : 제 3사분위 수 - 제 1사분위 수

In [38]:
df['평균 속도'].quantile(0.75) - df['평균 속도'].quantile(0.25)

14.854999999999997

## Questrion 18
읍면동명 컬럼의 유이값 갯수를 출력하라

In [49]:
df['읍면동명'].nunique()  # 유니크 갯수 출력 함수 : nunique()

41

#### Questrion 19
읍면동 칼럼의 유일값을 모두 출력하라

In [54]:
df['읍면동명'].unique()

array(['도두동', '외도동', '이도2동', '일도1동', '대천동', '서홍동', '한경면', '송산동', '조천읍',
       '일도2동', '영천동', '예래동', '대륜동', '삼도1동', '이호동', '건입동', '중앙동', '삼양동',
       '삼도2동', '이도1동', '남원읍', '대정읍', '정방동', '효돈동', '아라동', '한림읍', '구좌읍',
       '용담1동', '오라동', '화북동', '연동', '표선면', '중문동', '성산읍', '안덕면', '천지동',
       '노형동', '동홍동', '용담2동', '봉개동', '애월읍'], dtype=object)

# 02 Filtering & Sorting

#### Questrion 20
데이터를 로드하라

In [33]:
import pandas as pd
DataUrl = 'https://raw.githubusercontent.com/Datamanim/pandas/main/chipo.csv'
df = pd.read_csv(DataUrl)
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

#### Questrion 21
quantity 칼럼 값이 3인 데이터를 추출하여 첫 5행을 출력하라

In [34]:
df_quantity5 = df[df['quantity']==3]
df_quantity5.head(5)

order_id  quantity          item_name  \
409       178         3       Chicken Bowl   
445       193         3               Bowl   
689       284         3  Canned Soft Drink   
818       338         3      Bottled Water   
850       350         3  Canned Soft Drink   

                                    choice_description item_price  
409  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...    $32.94   
445  [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20   
689                                        [Diet Coke]     $3.75   
818                                                NaN     $3.27   
850                                           [Sprite]     $3.75

## Questrion 22
quantity 칼럼 값이 3인데이터를 추출하여 index를 0부터 정렬하고 첫 5행을 출력하라

In [35]:
df_quantity5_index = df[df['quantity']==3].reset_index(drop=True)  #  drop=True 옵션!
df_quantity5_index.head(5)

order_id  quantity          item_name  \
0       178         3       Chicken Bowl   
1       193         3               Bowl   
2       284         3  Canned Soft Drink   
3       338         3      Bottled Water   
4       350         3  Canned Soft Drink   

                                  choice_description item_price  
0  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...    $32.94   
1  [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20   
2                                        [Diet Coke]     $3.75   
3                                                NaN     $3.27   
4                                           [Sprite]     $3.75

#### Question 23
quantity, item_price 두 개의 칼럼으로 구성된 새로운 데이터 프레임을 정의하라

In [36]:
df2 = df[['quantity', 'item_price']]
df2

quantity item_price
0            1     $2.39 
1            1     $3.39 
2            1     $3.39 
3            1     $2.39 
4            2    $16.98 
...        ...        ...
4617         1    $11.75 
4618         1    $11.75 
4619         1    $11.25 
4620         1     $8.75 
4621         1     $8.75 

[4622 rows x 2 columns]

## Question 24
item_price 칼럼의 달러 표시 문자를 제거하고 float 타입으로 저장하여 new_price 칼럼에 저장하라

In [37]:
df['new_price'] = df['item_price'].str[1:].astype('float')  # 시작하는 $문자를 슬라이싱으로 자르고, astype()으로 형변환
df['new_price']

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: new_price, Length: 4622, dtype: float64

#### Question 25
new_price 칼럼이 5이하의 값을 가지는 데이터 프레임을 추출하고, 전체 갯수를 구하여라

In [38]:
len(df[df['new_price']<=5])

1652

## Question 26
item_name이 Chicken Salad Bowl 인 데이터 프레임을 추출하고 index 값을 초기화 하여라

In [39]:
df_item_name_Chiken= df[df['item_name']=='Chicken Salad Bowl'].reset_index(drop=True)  # drop=True 옵션과 inpace=Ture 옵션 헷갈림 ㅠㅠ
df_item_name_Chiken

order_id  quantity           item_name  \
0          20         1  Chicken Salad Bowl   
1          60         2  Chicken Salad Bowl   
2          94         2  Chicken Salad Bowl   
3         111         1  Chicken Salad Bowl   
4         137         2  Chicken Salad Bowl   
..        ...       ...                 ...   
105      1813         2  Chicken Salad Bowl   
106      1822         1  Chicken Salad Bowl   
107      1834         1  Chicken Salad Bowl   
108      1834         1  Chicken Salad Bowl   
109      1834         1  Chicken Salad Bowl   

                                    choice_description item_price  new_price  
0    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  
1    [Tomatillo Green Chili Salsa, [Sour Cream, Che...    $22.50       22.50  
2    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $22.50       22.50  
3    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75        8.75  
4              [Fresh Tomato Salsa, Fajita Vegetables]    $17.50       17.50  
..                                                 ...        ...        ...  
105  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $17.50       17.50  
106  [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...    $11.25       11.25  
107  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
108  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
109  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[110 rows x 6 columns]

#### Question 27
new_price 값이 9이하이고 item_name 값이 Chicken Salad Bowl 인 데이터 프레임을 추출하라

In [40]:
df4 = df[(df['new_price']<=9) & (df['item_name']=='Chicken Salad Bowl')]  # 반드시 소괄호 해주기!!
df4

order_id  quantity           item_name  \
44          20         1  Chicken Salad Bowl   
256        111         1  Chicken Salad Bowl   
526        220         1  Chicken Salad Bowl   
528        221         1  Chicken Salad Bowl   
529        221         1  Chicken Salad Bowl   
568        234         1  Chicken Salad Bowl   
585        240         1  Chicken Salad Bowl   
693        286         1  Chicken Salad Bowl   
702        289         1  Chicken Salad Bowl   
762        314         1  Chicken Salad Bowl   
1109       458         1  Chicken Salad Bowl   
1460       592         1  Chicken Salad Bowl   
1471       598         1  Chicken Salad Bowl   
1483       602         1  Chicken Salad Bowl   
1547       628         1  Chicken Salad Bowl   
1601       649         1  Chicken Salad Bowl   
1621       654         1  Chicken Salad Bowl   
1756       709         1  Chicken Salad Bowl   
1977       798         1  Chicken Salad Bowl   
2031       818         1  Chicken Salad Bowl   
2113       850         1  Chicken Salad Bowl   
2224       896         1  Chicken Salad Bowl   
2314       926         1  Chicken Salad Bowl   
2728      1083         1  Chicken Salad Bowl   
2858      1136         1  Chicken Salad Bowl   
2883      1145         1  Chicken Salad Bowl   
2897      1152         1  Chicken Salad Bowl   
2989      1189         1  Chicken Salad Bowl   
3033      1206         1  Chicken Salad Bowl   
3069      1222         1  Chicken Salad Bowl   
3117      1243         1  Chicken Salad Bowl   
3259      1305         1  Chicken Salad Bowl   
3261      1305         1  Chicken Salad Bowl   
3310      1327         1  Chicken Salad Bowl   
3315      1329         1  Chicken Salad Bowl   
3381      1357         1  Chicken Salad Bowl   
3411      1368         1  Chicken Salad Bowl   
3623      1450         1  Chicken Salad Bowl   
3656      1461         1  Chicken Salad Bowl   
3706      1483         1  Chicken Salad Bowl   
3712      1483         1  Chicken Salad Bowl   
3790      1517         1  Chicken Salad Bowl   
3798      1521         1  Chicken Salad Bowl   
3974      1592         1  Chicken Salad Bowl   
3997      1601         1  Chicken Salad Bowl   
4012      1606         1  Chicken Salad Bowl   
4035      1615         1  Chicken Salad Bowl   
4064      1627         1  Chicken Salad Bowl   
4205      1679         1  Chicken Salad Bowl   
4207      1679         1  Chicken Salad Bowl   
4333      1730         1  Chicken Salad Bowl   
4352      1738         1  Chicken Salad Bowl   
4439      1769         1  Chicken Salad Bowl   
4520      1797         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  new_price  
44    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  
256   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75        8.75  
526   [Roasted Chili Corn Salsa, [Black Beans, Sour ...     $8.75        8.75  
528   [Tomatillo Green Chili Salsa, [Fajita Vegetabl...     $8.75        8.75  
529   [Tomatillo Green Chili Salsa, [Fajita Vegetabl...     $8.75        8.75  
568             [Fresh Tomato Salsa, Fajita Vegetables]     $8.75        8.75  
585             [Fresh Tomato Salsa, Fajita Vegetables]     $8.75        8.75  
693   [Fresh Tomato Salsa, [Sour Cream, Cheese, Lett...     $8.75        8.75  
702   [Tomatillo Red Chili Salsa, [Black Beans, Chee...     $8.75        8.75  
762   [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
1109  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75        8.75  
1460                      [Fresh Tomato Salsa, Lettuce]     $8.75        8.75  
1471  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75        8.75  
1483  [Fresh Tomato Salsa, [Fajita Vegetables, Chees...     $8.75        8.75  
1547  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75    

## Question 28
df의 new_price 칼럼 값에 따라 오름차순으로 정리하고 index를 초기화하라

In [41]:
df.sort_values(by='new_price', ascending=True).reset_index(drop=True)

order_id  quantity                     item_name  \
0          471         1                 Bottled Water   
1          338         1                   Canned Soda   
2         1575         1                   Canned Soda   
3           47         1                   Canned Soda   
4         1014         1                   Canned Soda   
...        ...       ...                           ...   
4617      1443         3                Veggie Burrito   
4618      1443         4               Chicken Burrito   
4619       511         4               Chicken Burrito   
4620      1398         3                 Carnitas Bowl   
4621      1443        15  Chips and Fresh Tomato Salsa   

                                     choice_description item_price  new_price  
0                                                   NaN     $1.09        1.09  
1                                           [Coca Cola]     $1.09        1.09  
2                                          [Dr. Pepper]     $1.09        1.09  
3                                          [Dr. Pepper]     $1.09        1.09  
4                                           [Coca Cola]     $1.09        1.09  
...                                                 ...        ...        ...  
4617  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $33.75       33.75  
4618  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...    $35.00       35.00  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $35.00       35.00  
4620  [Roasted Chili Corn Salsa, [Fajita Vegetables,...    $35.25       35.25  
4621                                                NaN    $44.25       44.25  

[4622 rows x 6 columns]

## Question 29
df의 item_name 칼럼 값 중 Chips를 포함하는 경우의 데이터를 출력하라

In [42]:
# contains('포함할str')
df5 = df[df['item_name'].str.contains('Chips')]
df5

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
6            3         1                          Side of Chips   
10           5         1                    Chips and Guacamole   
14           7         1                    Chips and Guacamole   
...        ...       ...                                    ...   
4596      1826         1                    Chips and Guacamole   
4600      1827         1                    Chips and Guacamole   
4605      1828         1                    Chips and Guacamole   
4613      1831         1                                  Chips   
4616      1832         1                    Chips and Guacamole   

     choice_description item_price  new_price  
0                   NaN     $2.39        2.39  
3                   NaN     $2.39        2.39  
6                   NaN     $1.69        1.69  
10                  NaN     $4.45        4.45  
14                  NaN     $4.45        4.45  
...                 ...        ...        ...  
4596                NaN     $4.45        4.45  
4600                NaN     $4.45        4.45  
4605                NaN     $4.45        4.45  
4613                NaN     $2.15        2.15  
4616                NaN     $4.45        4.45  

[1084 rows x 6 columns]

## Question 30
df의 짝수번재 칼럼만을 포함하는 데이터프레임을 출력하라

In [43]:
df.iloc[:,::2]  # 컴마로 구분하지 않고 바로 :로 갖다 붙임

order_id                              item_name item_price
0            1           Chips and Fresh Tomato Salsa     $2.39 
1            1                                   Izze     $3.39 
2            1                       Nantucket Nectar     $3.39 
3            1  Chips and Tomatillo-Green Chili Salsa     $2.39 
4            2                           Chicken Bowl    $16.98 
...        ...                                    ...        ...
4617      1833                          Steak Burrito    $11.75 
4618      1833                          Steak Burrito    $11.75 
4619      1834                     Chicken Salad Bowl    $11.25 
4620      1834                     Chicken Salad Bowl     $8.75 
4621      1834                     Chicken Salad Bowl     $8.75 

[4622 rows x 3 columns]

#### Question 31
df의 new_price 칼럼 값에 따라 내림차순으로 정리하고 index를 초기화하라

In [44]:
df.sort_values(by='new_price', ascending=False).reset_index(drop=True)

order_id  quantity                     item_name  \
0         1443        15  Chips and Fresh Tomato Salsa   
1         1398         3                 Carnitas Bowl   
2          511         4               Chicken Burrito   
3         1443         4               Chicken Burrito   
4         1443         3                Veggie Burrito   
...        ...       ...                           ...   
4617      1578         1                   Canned Soda   
4618      1162         1                 Bottled Water   
4619       567         1                   Canned Soda   
4620      1014         1                   Canned Soda   
4621       591         1                   Canned Soda   

                                     choice_description item_price  new_price  
0                                                   NaN    $44.25       44.25  
1     [Roasted Chili Corn Salsa, [Fajita Vegetables,...    $35.25       35.25  
2     [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $35.00       35.00  
3     [Fresh Tomato Salsa, [Rice, Black Beans, Chees...    $35.00       35.00  
4     [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $33.75       33.75  
...                                                 ...        ...        ...  
4617                                  [Diet Dr. Pepper]     $1.09        1.09  
4618                                                NaN     $1.09        1.09  
4619                                        [Coca Cola]     $1.09        1.09  
4620                                        [Coca Cola]     $1.09        1.09  
4621                                           [Sprite]     $1.09        1.09  

[4622 rows x 6 columns]

#### Question 32
df의 item_name 칼럼 값이 Steak Salad 또는 Bowl 인 데이터를 인덱싱하라 (불리언 인덱싱 말하는 듯)

In [45]:
df[(df['item_name']=='Steak Salad') | (df['item_name']=='Bowl')]

order_id  quantity    item_name  \
445        193         3         Bowl   
664        276         1  Steak Salad   
673        279         1         Bowl   
752        311         1  Steak Salad   
893        369         1  Steak Salad   
3502      1406         1  Steak Salad   

                                     choice_description item_price  new_price  
445   [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20       22.20  
664   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99        8.99  
673   [Adobo-Marinated and Grilled Steak, [Sour Crea...     $7.40        7.40  
752   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99        8.99  
893   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...     $8.99        8.99  
3502                        [[Lettuce, Fajita Veggies]]     $8.69        8.69

## Question 33
df의 item_name 칼럼 값이 Steak Salad또는 Bowl인 데이터를 데이터프레임화 한 후,   
item_name을 기준으로 중복행이 있으면 제거하되 첫번째 케이스만 남겨라

In [46]:
df7 = df[(df['item_name']=='Steak Salad') | (df['item_name']=='Bowl')]
df7.drop_duplicates('item_name')  # drop_duplicates('기준이되는 칼럼') : 기준이 되는 칼럼을 기준으로 중복행 제거 메소드

order_id  quantity    item_name  \
445       193         3         Bowl   
664       276         1  Steak Salad   

                                    choice_description item_price  new_price  
445  [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20       22.20  
664  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99        8.99

#### Question 34
df의 item_name 칼럼 값이 Steak Salad또는 Bowl인 데이터를 데이터프레임화 한 후,   
item_name을 기준으로 중복행이 있으면 제거하되 마지막 케이스만 남겨라

In [47]:
df8 = df[(df['item_name']=='Steak Salad') | (df['item_name']=='Bowl')]
df8.drop_duplicates('item_name', keep='last')  # keep = 'last' 옵션!!

order_id  quantity    item_name  \
673        279         1         Bowl   
3502      1406         1  Steak Salad   

                                     choice_description item_price  new_price  
673   [Adobo-Marinated and Grilled Steak, [Sour Crea...     $7.40        7.40  
3502                        [[Lettuce, Fajita Veggies]]     $8.69        8.69

#### Question 35
df의 데이터 중 new_price값이 new_price값의 평균 이상을 가지는 데이터들을 인덱싱하라

In [52]:
import numpy as np
df[df['new_price'] >= np.mean(df['new_price'])]
# df.loc[df.new_price >= df.new_price.mean()]

order_id  quantity           item_name  \
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
8            4         1    Steak Soft Tacos   
9            5         1       Steak Burrito   
...        ...       ...                 ...   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  new_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98       10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75       11.75  
8     [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25        9.25  
9     [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25        9.25  
...                                                 ...        ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[2890 rows x 6 columns]

## Question 36
df의 데이터 중 item_name의 값의 lzze 데이터를 Fizzy Lizzy로 수정하라

In [62]:
df.loc[df['item_name']=='Izze', 'item_name'] = 'Fizzy Lizzy'  # 특정 조건의 데이터를 바꿀 때 바꿀 데이터의 컬럼을 명시해줘야한다.
df.head(3)

order_id  quantity                     item_name choice_description  \
0         1         1  Chips and Fresh Tomato Salsa                NaN   
1         1         1                   Fizzy Lizzy       [Clementine]   
2         1         1              Nantucket Nectar            [Apple]   

  item_price  new_price  
0     $2.39        2.39  
1     $3.39        3.39  
2     $3.39        3.39

#### Question 37
df의 데이터 중 choice_description 값이 NaN인 데이터의 갯수를 구하여라

In [69]:
# len(df[df['choice_description'].isnull() == True])
df['choice_description'].isnull().sum()

1246

#### Question 38
df의 데이터 중 choice_description 값이 NaN인 데이터를 NoData 값으로 대체하라(loc 이용)

In [70]:
df.loc[df.choice_description.isnull()==True, 'choice_description'] = 'NoData'
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                            Fizzy Lizzy   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  new_price  
0                                             NoData     $2.39        2.39  
1                                       [Clementine]     $3.39        3.39  
2                                            [Apple]     $3.39        3.39  
3                                             NoData     $2.39        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98

#### Question 39
df의 데이터 중 choice_description 값에 Black이 들어가는 경우를 인덱싱하라

In [72]:
df.loc[df.choice_description.str.contains('Black')==True]

order_id  quantity             item_name  \
4            2         2          Chicken Bowl   
7            4         1         Steak Burrito   
9            5         1         Steak Burrito   
11           6         1  Chicken Crispy Tacos   
12           6         1    Chicken Soft Tacos   
...        ...       ...                   ...   
4604      1828         1          Chicken Bowl   
4608      1829         1        Veggie Burrito   
4611      1830         1        Veggie Burrito   
4612      1831         1         Carnitas Bowl   
4617      1833         1         Steak Burrito   

                                     choice_description item_price  new_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75       11.75  
9     [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25        9.25  
11    [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75        8.75  
12    [Roasted Chili Corn Salsa, [Rice, Black Beans,...     $8.75        8.75  
...                                                 ...        ...        ...  
4604  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...     $8.75        8.75  
4608  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.25       11.25  
4611  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...    $11.25       11.25  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25        9.25  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  

[1353 rows x 6 columns]

#### Question 40
df의 데이터 중 choice_description 값에 Vegetables 들어가지 않는 경우의 갯수를 출력하라

In [81]:
len(df.loc[df.choice_description.str.contains('Vegetables')==False])

3900

#### Question 41
df의 데이터 중 item_name 값이 N으로 시작하는 데이터를 모두 추출하라

In [82]:
df.loc[df.item_name.str.startswith('N')]

order_id  quantity         item_name         choice_description  \
2            1         1  Nantucket Nectar                    [Apple]   
22          11         1  Nantucket Nectar       [Pomegranate Cherry]   
105         46         1  Nantucket Nectar  [Pineapple Orange Banana]   
173         77         1  Nantucket Nectar                    [Apple]   
205         91         1  Nantucket Nectar             [Peach Orange]   
436        189         1  Nantucket Nectar       [Pomegranate Cherry]   
601        247         2  Nantucket Nectar  [Pineapple Orange Banana]   
925        381         1  Nantucket Nectar       [Pomegranate Cherry]   
1356       553         1  Nantucket Nectar       [Pomegranate Cherry]   
1585       641         1  Nantucket Nectar             [Peach Orange]   
1626       656         1  Nantucket Nectar  [Pineapple Orange Banana]   
1706       690         1  Nantucket Nectar                    [Apple]   
2162       872         1  Nantucket Nectar  [Pineapple Orange Banana]   
2379       947         2  Nantucket Nectar             [Peach Orange]   
2381       947         1  Nantucket Nectar                    [Apple]   
2430       965         1  Nantucket Nectar       [Pomegranate Cherry]   
2653      1053         1  Nantucket Nectar  [Pineapple Orange Banana]   
2818      1118         1  Nantucket Nectar                    [Apple]   
2838      1128         1  Nantucket Nectar             [Peach Orange]   
2853      1133         1  Nantucket Nectar                    [Apple]   
2949      1172         1  Nantucket Nectar             [Peach Orange]   
3318      1330         1  Nantucket Nectar             [Peach Orange]   
3368      1351         1  Nantucket Nectar  [Pineapple Orange Banana]   
3570      1433         1  Nantucket Nectar  [Pineapple Orange Banana]   
3845      1541         1  Nantucket Nectar             [Peach Orange]   
4019      1609         1  Nantucket Nectar  [Pineapple Orange Banana]   
4078      1632         1  Nantucket Nectar             [Peach Orange]   

     item_price  new_price  
2        $3.39        3.39  
22       $3.39        3.39  
105      $3.39        3.39  
173      $3.39        3.39  
205      $3.39        3.39  
436      $3.39        3.39  
601      $6.78        6.78  
925      $3.39        3.39  
1356     $3.39        3.39  
1585     $3.39        3.39  
1626     $3.39        3.39  
1706     $3.39        3.39  
2162     $3.39        3.39  
2379     $6.78        6.78  
2381     $3.39        3.39  
2430     $3.39        3.39  
2653     $3.39        3.39  
2818     $3.39        3.39  
2838     $3.39        3.39  
2853     $3.39        3.39  
2949     $3.39        3.39  
3318     $3.39        3.39  
3368     $3.39        3.39  
3570     $3.39        3.39  
3845     $3.39        3.39  
4019     $3.39        3.39  
4078     $3.39        3.39

#### Question 42
df의 데이터 중 item_name 값의 단어갯수가 15개 이상인 데이터를 인덱싱하라

In [84]:
df.loc[df.item_name.str.len() >=15]

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
8            4         1                       Steak Soft Tacos   
10           5         1                    Chips and Guacamole   
...        ...       ...                                    ...   
4615      1832         1                     Chicken Soft Tacos   
4616      1832         1                    Chips and Guacamole   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  new_price  
0                                                NoData     $2.39        2.39  
2                                               [Apple]     $3.39        3.39  
3                                                NoData     $2.39        2.39  
8     [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25        9.25  
10                                               NoData     $4.45        4.45  
...                                                 ...        ...        ...  
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75        8.75  
4616                                             NoData     $4.45        4.45  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[2373 rows x 6 columns]

## Question 43
df의 데이터 중 new_price값이 lst에 해당하는 경우의 데이터 프레임을 구하고 그 갯수를 출력하라   
lst=[1.69, 2.39, 3.39, 4.45, 9.25, 10.98, 11.75, 16.98]

In [88]:
lst=[16.9, 2.39, 3.39, 4.45, 9.25, 10.98, 11.75, 16.98]
df.loc[df.new_price.isin(lst)]  # isin() 앞에는 str을 붙이지 않는다. in이 아니라 isin인 것도 기억!

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                            Fizzy Lizzy   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4610      1830         1                          Steak Burrito   
4612      1831         1                          Carnitas Bowl   
4616      1832         1                    Chips and Guacamole   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   

                                     choice_description item_price  new_price  
0                                                NoData     $2.39        2.39  
1                                          [Clementine]     $3.39        3.39  
2                                               [Apple]     $3.39        3.39  
3                                                NoData     $2.39        2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
...                                                 ...        ...        ...  
4610  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25        9.25  
4616                                             NoData     $4.45        4.45  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  

[1294 rows x 6 columns]

# 3_Grouping

#### Question 44
데이터를 로드하고 상위 5개 컬럼을 출력하라

In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/AB_NYC_2019.csv')
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## Question 45
데이터의 각 host_name의 빈도수를 구하고 host_name으로 정렬하여 상위 5개를 출력하라

In [2]:
# groupby 심화
## 판다스 groupby()의 기능은 데이터를 그룹별로 분할하여 독립된 그룹에 대하여 별도로 데이터를 처리(혹은 적용)하거나  
## 그룹별 통계량을 확인하고자 할 때 유용한 함수이다. 
### 1) Split 단계 : .groupby()에서 정의한 칼럼 조건에 따라 독립된 그룹으로 나누어준다. 
### 2) Apply 단계 : Split단계에서 나누어진 독립된 그룹별 원하는 함수를 적용하는 단계이다.
### 3) Combine 단걔 : 각각의 독립된 그룹별로 함수가 적용된 결과를 종합하여 다시 하나의 테이블로 합치는 단계이다.

df0 = df.groupby('host_name').size()
df0.head(5)

### 여기서는 'host_name'을 기준으로 split, size()함수를 적용, 다시 합쳐서 내놓은 결과를 상위 5개 보여준다.

host_name
'Cil                        1
(Ari) HENRY LEE             1
(Email hidden by Airbnb)    6
(Mary) Haiy                 1
-TheQueensCornerLot         1
dtype: int64

## Question 46
데이터의 각 host_name의 빈도수를 구하고 빈도수 기준 내림차순 정렬한 데이터 프레임을 만들어라.   
빈도수 컬럼은 counts로 명명하라

In [3]:
df1 = df.groupby('host_name').size().sort_values(0, ascending=False)
df1.to_frame().rename(columns={0:'counts'})

# .groupby를 이용해 데이터 프레임을 새로 만들면 그룹의 기준이 되는게 인덱스로 들어가고 함수 수행 결과 합쳐진 값이 칼럼 0으로 들어간다.

/var/folders/ps/szc1qwh135x62ct9q9t4phnr0000gn/T/ipykernel_966/3186509876.py:1: FutureWarning: In a future version of pandas all arguments of Series.sort_values will be keyword-only.
  df1 = df.groupby('host_name').size().sort_values(0, ascending=False)


counts
host_name           
Michael          417
David            403
Sonder (NYC)     327
John             294
Alex             279
...              ...
Jerbean            1
Jerald             1
Jeonghoon          1
Jeny               1
현선                 1

[11452 rows x 1 columns]

In [4]:
# 특정 칼럼의 빈도수 구하는 메소드 .groupby('특정칼럼').size()
df.groupby('host_name').size()\
                       .to_frame().rename(columns={0:'counts'}) \
                       .sort_values('counts', ascending=False) 

counts
host_name           
Michael          417
David            403
Sonder (NYC)     327
John             294
Alex             279
...              ...
Jerbean            1
Jerald             1
Jeonghoon          1
Jeny               1
현선                 1

[11452 rows x 1 columns]

## Question 47
neighbourhood_group의 값에 따른 neighbourhood컬럼 값의 갯수를 구하여라

In [5]:
# as_index=False 옵션을 주면 그룹바이 기준이 되었던 컬럼이 인덱스에 자리잡지 않게 된다.
# 그리고 그룹바이 기준을 다중으로할 경우 리스트 형태로 넣어줘야 한다.
df2 = df.groupby(['neighbourhood_group', 'neighbourhood'], as_index=False).size()

## Question 48
neighbourhood_group의 값에 따른 neighbourhood컬럼 값 중 neighbourhood_group그룹의 최댓값들을 출력하라

In [6]:
df.groupby(['neighbourhood_group', 'neighbourhood'], as_index=False).size() \
  .groupby(['neighbourhood_group'], as_index=False).max()

neighbourhood_group    neighbourhood  size
0               Bronx         Woodlawn    70
1            Brooklyn  Windsor Terrace  3920
2           Manhattan     West Village  2658
3              Queens         Woodside   900
4       Staten Island          Woodrow    48

In [7]:
# 위의 코드에서 groupby()를 두번 연속 사용하는 이유는 데이터를 걔층적으로 그룹화하기 위해서이다.

# 첫번째 gropuby()에서는 'neighbourhood_group'을 기준으로 그룹화 한다. 
# 결과적으로 각 'neighbourhood_group'와 'neighbourhood'별로 데이터가 그룹화 된다.
# 그런 다음 size()를 사용하여 각 그룹 내의 관측치 수를 계산한다.

# 두번째 groupby()에서는 'neighbourhood_group'을 기준으로 그룹화한다.
# 그룹화된 결과를 다시 그룹화하면, 'neighbourhood_group'별로 size로 구한 관측치에 함수를 적용할 수 있다.

# 이러한 방식으로 두 번의 groupby()를 사용하면 계층적으로 그룹화된 데이터에서 원하는 결과를얻을 수 있다.
# 즉, 최종 결과는 'neighbourhood_group'별로 관측치 수의 최댓값을 포함하는 데이터프레임이다.고

## Question 49
neighbourhood_group 값에 따른 price값의 평균, 분산, 최대, 최소 값을 구하여라

In [8]:
df3 = df[['neighbourhood_group', 'price']].groupby('neighbourhood_group').agg(['mean', 'var', 'max', 'min'])
df3

# agg([var:분산, mean:평균 등]) 으로 여러개의 집계함수 적용 가능

price                         
                           mean           var    max min
neighbourhood_group                                     
Bronx                 87.496792  11386.885081   2500   0
Brooklyn             124.383207  34921.719135  10000   0
Manhattan            196.875814  84904.159185  10000   0
Queens                99.517649  27923.130227  10000  10
Staten Island        114.812332  77073.088342   5000  13

#### Question 50
neighbourhood_group 값에 따른 reviews_per_month 평균, 분산, 최대, 최소 값을 구하여라

In [9]:
df[['neighbourhood_group', 'reviews_per_month']].groupby('neighbourhood_group').agg(['mean', 'var', 'max', 'min'])

reviews_per_month                       
                                 mean       var    max   min
neighbourhood_group                                         
Bronx                        1.837831  2.799878  10.34  0.02
Brooklyn                     1.283212  2.299040  14.00  0.01
Manhattan                    1.272131  2.651206  58.50  0.01
Queens                       1.941200  4.897848  20.94  0.01
Staten Island                1.872580  2.840895  10.12  0.02

## Question 51
neighbourhood 값과 neighbourhood_group 값에 따른 price의 평균을 구하라

In [15]:
df.groupby(['neighbourhood_group', 'neighbourhood']).price.mean()

# 우리가 원하는 최종 칼럼 값은 price의 평균이지만 groupby에는 price를 넣지 않는다.
# price를 기준으로 그룹을 형성하지 않기 때문이다.
# 단지 기준으로 삼을 그룹들로 묶어주고 .price 를 사용한다.

neighbourhood_group  neighbourhood
Bronx                Allerton          87.595238
                     Baychester        75.428571
                     Belmont           77.125000
                     Bronxdale         57.105263
                     Castle Hill       63.000000
                                         ...    
Staten Island        Tottenville      144.857143
                     West Brighton     80.555556
                     Westerleigh       71.500000
                     Willowbrook      249.000000
                     Woodrow          700.000000
Name: price, Length: 221, dtype: float64

#### Question 52
neighbourhood 값과 neighbourhood_group 값에 따른 price의 평균을 계층적 indexing 없이 구하라

In [18]:
df.groupby(['neighbourhood', 'neighbourhood_group']).price.mean().unstack()

# 계층적 indexing 없이 < 이 말 뜻을 알아야 한다. 이것은 unstack() 메소드를 사용하라는 뜻이다.
# unstack() : 다중 인덱스가 있는 데이터프레임에서 인덱스의 레벨을 변환하여 데이터의 구조를 재조정하는 역할을 한다.
# groupby()안에 컬럼 리스트의 순서를 앞뒤로 바꾼 경우 출력되는 구조가 다르기 때문에 주의가 필요하다.

neighbourhood_group      Bronx    Brooklyn  Manhattan      Queens  \
neighbourhood                                                       
Allerton             87.595238         NaN        NaN         NaN   
Arden Heights              NaN         NaN        NaN         NaN   
Arrochar                   NaN         NaN        NaN         NaN   
Arverne                    NaN         NaN        NaN  171.779221   
Astoria                    NaN         NaN        NaN  117.187778   
...                        ...         ...        ...         ...   
Windsor Terrace            NaN  138.993631        NaN         NaN   
Woodhaven                  NaN         NaN        NaN   67.170455   
Woodlawn             60.090909         NaN        NaN         NaN   
Woodrow                    NaN         NaN        NaN         NaN   
Woodside                   NaN         NaN        NaN   85.097872   

neighbourhood_group  Staten Island  
neighbourhood                       
Allerton                       NaN  
Arden Heights                67.25  
Arrochar                    115.00  
Arverne                        NaN  
Astoria                        NaN  
...                            ...  
Windsor Terrace                NaN  
Woodhaven                      NaN  
Woodlawn                       NaN  
Woodrow                     700.00  
Woodside                       NaN  

[221 rows x 5 columns]

## Question 53
neighbourhood 값과 neighbourhood_group 값에 따른 price의 평균을 계층적 indexing 없이 구하고 nan 값은 -999값으로 채워라

In [21]:
df.groupby(['neighbourhood', 'neighbourhood_group']).price.mean().unstack().fillna(-999)

# NaN 값을 다른 값으로 대체하는 메소드 : fillna()

neighbourhood_group       Bronx    Brooklyn  Manhattan      Queens  \
neighbourhood                                                        
Allerton              87.595238 -999.000000     -999.0 -999.000000   
Arden Heights       -999.000000 -999.000000     -999.0 -999.000000   
Arrochar            -999.000000 -999.000000     -999.0 -999.000000   
Arverne             -999.000000 -999.000000     -999.0  171.779221   
Astoria             -999.000000 -999.000000     -999.0  117.187778   
...                         ...         ...        ...         ...   
Windsor Terrace     -999.000000  138.993631     -999.0 -999.000000   
Woodhaven           -999.000000 -999.000000     -999.0   67.170455   
Woodlawn              60.090909 -999.000000     -999.0 -999.000000   
Woodrow             -999.000000 -999.000000     -999.0 -999.000000   
Woodside            -999.000000 -999.000000     -999.0   85.097872   

neighbourhood_group  Staten Island  
neighbourhood                       
Allerton                   -999.00  
Arden Heights                67.25  
Arrochar                    115.00  
Arverne                    -999.00  
Astoria                    -999.00  
...                            ...  
Windsor Terrace            -999.00  
Woodhaven                  -999.00  
Woodlawn                   -999.00  
Woodrow                     700.00  
Woodside                   -999.00  

[221 rows x 5 columns]

## Question 54
데이터중 neighbourhood_group 값이 Queens값을 가지는 데이터들 중 neighbourhood 그룹별로  
price값의 평균, 분산, 최대. 최소값을 구하라

In [26]:
df5 = df.loc[df.neighbourhood_group == 'Queens']
df5.groupby('neighbourhood').price.agg(['mean', 'var', 'max', 'min'])

# agg() 메소드를 arr()로 헛갈리지 말자 ㅠㅠ

mean            var    max  min
neighbourhood                                             
Arverne              171.779221   37383.411141   1500   35
Astoria              117.187778  122428.811196  10000   25
Bay Terrace          142.000000    6816.400000    258   32
Bayside              157.948718  166106.470985   2600   30
Bayswater             87.470588    2330.889706    230   45
Belle Harbor         171.500000    8226.571429    350   85
Bellerose             99.357143    3093.016484    240   42
Breezy Point         213.333333    1008.333333    250  195
Briarwood            105.875000   18503.165909   1000   30
Cambria Heights       81.730769    2960.604615    250   31
College Point         88.000000    6445.555556    400   44
Corona                59.171875    2430.557292    359   23
Ditmars Steinway      95.029126    4214.112785    600   22
Douglaston            82.750000    2347.071429    178   40
East Elmhurst         81.183784    4369.487779    700   16
Edgemere              94.727273    4219.418182    200   40
Elmhurst              80.459916    3896.563005    443   15
Far Rockaway         165.862069   52575.980296    900   35
Flushing              93.514085   12417.177448   1500   15
Forest Hills         121.625000   44533.368881   2350   16
Fresh Meadows         99.500000    6394.451613    375   39
Glendale              90.796296    3230.882250    299   25
Hollis                88.642857    1483.016484    175   50
Holliswood           135.750000    5091.583333    239   79
Howard Beach         115.400000    4241.936842    250   40
Jackson Heights       80.897849    2150.470590    260   23
Jamaica               95.770563    8067.047130   1000   10
Jamaica Estates      182.947368   29006.052632    750   35
Jamaica Hills        132.125000   11304.696429    325   50
Kew Gardens           88.375000    2632.306452    219   35
Kew Gardens Hills    112.307692    6230.621538    399   40
Laurelton             95.333333    4355.529412    254   34
Little Neck           75.200000    2368.700000    149   32
Long Island City     127.465549   15529.503009   2000   30
Maspeth               83.645455    2223.607089    245   11
Middle Village       109.580645    3537.118280    265   28
Neponsit             274.666667    5625.333333    350  200
Ozone Park            85.274194    1705.415389    200   25
Queens Village        83.933333    2386.470056    320   25
Rego Park             83.877358    3659.003863    300   21
Richmond Hill         87.117021    3474.147449    300   28
Ridgewood             77.184397    2629.311889    375   20
Rockaway Beach       132.178571    6254.622078    545   49
Rosedale              76.694915    2963.939801    350   22
South Ozone Park      82.400000    4619.989744    400   29
Springfield Gardens   94.235294    2490.848739    300   40
St. Albans           100.828947    9489.903684    600   25
Sunnyside             84.865014    2727.746922    600   12
Whitestone           107.545455   13632.072727    400   35
Woodhaven             67.170455    1642.464864    250   10
Woodside              85.097872    5001.900636    500   28

## Question 55
데이터중 neighbourhood_group 값에 따른 room_type 컬럼의 숫자를 구하고  
neighbourhood_group 값을 기준으로 각 값의 비율을 구하여라 

In [43]:
# 값의 비율을 구하는 문제
sub1 = df.groupby(['neighbourhood_group', 'room_type']).room_type.size().unstack()
sub1.loc[:,:] = (sub1.values / sub1.sum(axis=1).values.reshape(-1,1))
# 두 번째 줄 코드는 계싼된 그룹 크기를 각 행의 총합으로 나누어 비율로 변환하는 작업이다. 
# sub1[:,:]를 통해 계싼된 비율로 데이터프레임의 값을 대체한다.
# axis=1은 각 행을 따라 연산을 수행하르는 의미이다.
# values.reshape(-1,1)은 행렬의 형태를 변경하여 열이 1개인 형태로 변환하는 것이다.

# reshape(-1,1)은 넘파이 배열의 형태를 변경하는 메소드이다.  
# -1은 해당 차원의 크기를 자동으로 계산하라는 의미이다.
# 괄호 앞에 행의 차원 수, 괄호 뒤에 열의 차원 수가 들어가는 여기서는 행은 자동으로 열은 1차원 이다.

sub1

room_type            Entire home/apt  Private room  Shared room
neighbourhood_group                                            
Bronx                       0.347388      0.597617     0.054995
Brooklyn                    0.475478      0.503979     0.020543
Manhattan                   0.609344      0.368496     0.022160
Queens                      0.369926      0.595129     0.034945
Staten Island               0.471850      0.504021     0.024129

# 04_Apply, Map 

#### Question 56
데이터를 로드하고 데이터 행과 열의 갯수를 출력하라

In [46]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/pandas/main/BankChurnersUp.csv', index_col=0)
df.shape

(10127, 18)

## Question 57
Income_Category의 카테고리를 map 함수를 이용하여 다음과 같이 변경하여 newIncome 컬럼에 매핑하라  
Unknown : N  
Less than $40K : a  
$40K - $60K : b  
$60K - $80K : c  
$80K - $120K : d  
$120 +' : e


In [53]:
# 칼럼에서 개개인의 값에 어떻게 접근할까? apply 또는 map을 사용할거 같긴함

# map 함수 사용해서 하는 방법
dic = {
    'Unknown'           : 'N',
    'Less than $40K'    : 'a',
    '$40K - $60K'       : 'b',  
    '$60K - $80K'       : 'c',
    '$80K - $120K'      : 'd', 
    '$120K +'            : 'e',
}
df['newIncome'] = df['Income_Category'].map(lambda x : dic[x])
df['newIncome']

# map과 lambda함수 그리고 딕셔너리를 이용해서 매핑할 수 있다.

# map은 딕셔너리로 키값에 해당하는 문자열을 밸류값으로 매핑할 수 있다.

0        c
1        a
2        d
3        a
4        c
        ..
10122    b
10123    b
10124    a
10125    b
10126    a
Name: newIncome, Length: 10127, dtype: object

#### Question 58
Income_Category의 카테고리를 apply 함수를 이용하여 다음과 같이 변경하여 newIncome 컬럼에 매핑하라   
Unknown : N  
Less than $40K : a  
$40K - $60K : b  
$60K - $80K : c  
$80K - $120K : d  
$120K +’ : e  

In [54]:
# apply함수 이용해서 하는 방법
# apply 메소드는 사용자 정의 함수를 각 밸류마다 적용시키는 것이다.

def changeValue(x):
    if x=='Unkown':
        return 'N'
    elif x=='Less than $40K':
        return 'a'
    elif x=='$40K - $60K':
        return 'b'
    elif x=='$60K - $80K':
        return 'c'
    elif x=='$80K - $120K':
        return 'd'
    elif x=='$120K +':
        return 'e'
    
df['newIncome'] = df.Income_Category.apply(changeValue)

df['newIncome']

0        c
1        a
2        d
3        a
4        c
        ..
10122    b
10123    b
10124    a
10125    b
10126    a
Name: newIncome, Length: 10127, dtype: object

#### Question 59
Customer_Age의 값을 이용하여 나이 구간을 AgeState 칼럼으로 정의하라.  
(0~9 : 0 , 10~19 :10 , 20~29 :20 …) 각 구간의 빈도수를 출력하라


In [58]:
min(df['Customer_Age'])

26

In [64]:
# 이건 apply 이용하는게 더 쉽겠다.

def to_ageStat(x):
    if x >= 0 and x < 10:
        return 0
    elif x < 20:
        return 10
    elif x < 30:
        return 20
    elif x < 40:
        return 30
    elif x < 50:
        return 40
    elif x < 60:
        return 50
    elif x < 70:
        return 60
    elif x < 80:
        return 70

df['AgeState'] = df['Customer_Age'].apply(to_ageStat)

df['AgeState'].value_counts().sort_index()
# .sort_index까지 해줘야 인덱스인 나이대 순으로 정렬 되어서 출력된다.
# .sort_index사용하지 않으면 컬럼인 빈도수가 높은 순으로 정렬 된다.

40    4561
50    2998
30    1841
60     530
20     195
70       2
Name: AgeState, dtype: int64

## Question 60
Education_Level의 값중 Graduate단어가 포함되는 값은 1 그렇지 않은 경우에는 0으로 변경하여   
newEduLevel 칼럼을 정의하고 빈도수를 출력하라

In [86]:
def aboutEL(x):
    if 'Graduate' in x :
        return 1
    else: 
        return 0

df['newEduLevel'] = df['Education_Level'].apply(aboutEL)
df['newEduLevel'].value_counts()

# 'Graduate'와 x의 위치가 in을 사이에 두고 어디에 있는지 매우 중요 (착각하면 전혀 다른 값이 나옴)
# value_counts()와 size()의 차이가 무엇?
# value_counts() : 시리즈에 사용되며 '고유한 값의 빈도수'를 내림차순 정렬된 시리즈로 반환
# size() : 데이터프레임이나 시리즈의 전체 크기(데이터 개수)를 반환하는 메소드. 즉, 행과 열에 포함된 모든 데이터의 개수를 반환

0    6483
1    3644
Name: newEduLevel, dtype: int64

#### Question 61
Credit_Limit 컬럼값이 4500 이상인 경우 1 그외의 경우에는 모두 0으로 하는 newLimit 정의하라.  
newLimit 각 값들의 빈도수를 출력하라

#### Question 62
Marital_Status 컬럼값이 Married 이고 Card_Category 칼럼의 값이 Platinum인 경우 1  
그외의 경우에는 모두 0으로 하는 newState칼럼을 정의하라. newState의 각 값들의 빈도수를 출력하라

#### Question 63
Gender 컬럼값 M인 경우 male, F인 경우 female로 값을 변경하여 Gender 컬럼에 새롭게 정의하라.  
각 value의 빈도를 출력하라

# 05_Time_Series

#### Question 64
데이터를 로드하고 각 열의 데이터 타입을 파악하라

#### Question 65
Yr_Mo_Dy을 판다스에서 인식할 수 있는 datetime64타입으로 변경하라

#### Question 66
Yr_Mo_Dy에 존재하는 년도의 유일값을 모두 출력하라

#### Question 67
Yr_Mo_Dy에 년도가 2061년 이상의 경우에는 모두 잘못된 데이터이다.  
해당경우의 값은 100을 빼서 새롭게 날짜를 Yr_Mo_Dy 컬럼에 정의하라

#### Question 68
년도별 각컬럼의 평균값을 구하여라

#### Question 69
weekday컬럼을 만들고 요일별로 매핑하라 (월요일:0 ~ 일요일:6)

#### Question 70
weekday칼럼을 기준으로 주말이면 1 평일이면 0의 값을 가지는 WeekCheck 칼럼을 만들어라

#### Question 71
년도, 일자 상관없이 모든 칼럼의 각 달의 평균을 구하여라

#### Question 72
모든 결측치는 컬럼기준 직전의 값으로 대체하고 첫번째 행에 결측치가 있을 경우 뒤에 있는 값으로 대체하라

#### Question 73
년도 -월을 기준으로 모든 칼럼의 평균값을 구하여라

#### Question 74
RPT 컬럼의 값을 일자별 기준으로 1차차분하라

#### Question 75
RPT와 VAL의 칼럼을 일주일 간격으로 각각 이동평균한값을 구하여라

#### Question 76
년-월-일:시 컬럼을 pandas에서 인식할 수 있는 datetime 형태로 변경하라. 서울시의 제공데이터의 경우 0시가 24시로 표현된다

#### Question 77
일자별 영어요일 이름을 dayName 컬럼에 저장하라

#### Question 78
일자별 각 PM10등급의 빈도수를 파악하라

#### Question 79
시간이 연속적으로 존재하며 결측치가 없는지 확인하라

#### Question 80
오전 10시와 오후 10시(22시)의 PM10의 평균값을 각각 구하여라

#### Question 81
날짜 컬럼을 index로 만들어라

#### Question 82
데이터를 주단위로 뽑아서 최소, 최대 평균, 표준편차를 구하여라

# 06_Pivot

#### Question 83
Indicator을 삭제하고 First Tooltip 칼럼에서 신뢰구간에 해당하는 표현을 지워라

#### Question 84
년도가 2015년 이상, Dim1이 Both sexes인 케이스만 추출하라

#### Question 85
84번 문제에서 추출한 데이터로 아래와 같이 나라에 따른 년도별 사망률을 데이터 프레임화 하라

#### Question 86
Dim1에 따른 년도별 사망비율의 평균을 구하라

#### Question 87
데이터에서 한국 KOR 데이터만 추출하라

#### Question 88
한국 올림픽 메달리스트 데이터에서 년도에 따른 medal 갯수를 데이터프레임화 하라

#### Question 89
전체 데이터에서 sport종류에 따른 성별수를 구하여라

#### Question 90
전체 데이터에서 Discipline종류에 따른 Medal수를 구하여라

# 07_Merge, Concat

#### Question 91
df1과 df2 데이터를 하나의 데이터 프레임으로 합쳐라

#### Question 92
df3과 df4 데이터를 하나의 데이터 프레임으로 합쳐라. 둘 다 포함하고 있는 년도에 대해서만 고려한다

#### Question 93
df3과 df4 데이터를 하나의 데이터 프레임으로 합쳐라. 모든 컬럼을 포함하고, 결측치는 0으로 대체한다

#### Question 94
df5과 df6 데이터를 하나의 데이터 프레임으로 merge함수를 이용하여 합쳐라.  
Algeria컬럼을 key로 하고 두 데이터 모두 포함하는 데이터만 출력하라

#### Question 95
df5과 df6 데이터를 하나의 데이터 프레임으로 merge함수를 이용하여 합쳐라.  
Algeria 컬럼을 key로 하고 합집합으로 합쳐라